utils/azureCredentialsManager.ts

```diff
@@ -135,4 +135,292 @@ export class AzureCredentialsManager {
+        const { accessToken, refreshToken } = await this.acquireToken(session);
+
+        //regenerates in case they have expired
+        if (accessToken && refreshToken) {
+            let refreshTokenARC;
+            let accessTokenARC;
+
+            await request.post('https://' + registry.loginServer + '/oauth2/exchange', {
+                form: {
+                    grant_type: 'access_token_refresh_token',
+                    service: registry.loginServer,
+                    tenant: tenantId,
+                    refresh_token: refreshToken,
+                    access_token: accessToken
+                }
+            }, (err, httpResponse, body) => {
+                if (body.length > 0) {
+                    refreshTokenARC = JSON.parse(body).refresh_token;
+                } else {
+                    return;
+                }
+            });
+
+            await request.post('https://' + registry.loginServer + '/oauth2/token', {
+                form: {
+                    grant_type: 'refresh_token',
+                    service: registry.loginServer,
+                    scope: 'registry:catalog:*',
+                    refresh_token: refreshTokenARC
+                }
+            }, (err, httpResponse, body) => {
+                if (body.length > 0) {
+                    accessTokenARC = JSON.parse(body).access_token;
+                } else {
+                    return;
+                }
+            });
+            if (refreshTokenARC && accessTokenARC) {
+                return { 'refreshToken': refreshTokenARC, 'accessToken': accessTokenARC };
+            }
+        }
+        return { refreshToken, accessToken }
+    }
+
+    public async acquireToken(localSession: AzureSession): Promise<{ accessToken: string; refreshToken: string; }> {
+        return new Promise<{ accessToken: string; refreshToken: string; }>((resolve, reject) => {
+            const credentials: any = localSession.credentials;
+            const environment: any = localSession.environment;
+            // tslint:disable-next-line:no-function-expression // Grandfathered in
+            credentials.context.acquireToken(environment.activeDirectoryResourceId, credentials.username, credentials.clientId, function (err: any, result: { accessToken: string; refreshToken: string; }): void {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve({
+                        accessToken: result.accessToken,
+                        refreshToken: result.refreshToken
+                    });
+                }
+            });
+        });
+    }
+
+    /**
+     *
+     * @param username : username for creating header
+     * @param password : password for creating header
+     */
+    public _get_authorization_header(username: string, password: string): string {
```

Comment:
This should not be public, additionally authorization header for? There can be many so make sure you scope this.